MAINT: add missing dunder method to nditer type hints

MatthieuDartiailh opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

Closes #19551

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -778,6 +779,21 @@ class nditer:
+    def __enter__(self) -> nditer: ...
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        traceback: Optional[TracebackType]
```
```suggestion
        exc_type: None | Type[BaseException],
        exc_value: None | BaseException,
        traceback: None | TracebackType,
```
Style nit

MatthieuDartiailh replied:
I agree for Union not so much  for Optional but I will make the change. Since PEP604 is 3.10 I would not have though to use it but I guess the constraint is lighter in type hints.

BvB93 replied:
> Since PEP604 is 3.10 I would not have though to use it but I guess the constraint is lighter in type hints.

Yup, the `|` syntax has been supported for annotations (not type aliases) since the [mypy 0.800 release](https://mypy-lang.blogspot.com/2021/01/mypy-0800-released.html).

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -778,6 +779,21 @@ class nditer:
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        traceback: Optional[TracebackType]
+    ) -> Optional[bool]: ...
```
```suggestion
    ) -> None: ...
```
`nditer.__exit__` always returns `None`

BvB93 commented on numpy/__init__.pyi:
```diff
@@ -778,6 +779,21 @@ class nditer:
+    def __next__(self) -> Any: ...
+    def __len__(self) -> int: ...
+    def __copy__(self) -> nditer: ...
+    def __getitem__(self, index: int) -> Any: ...
+    def __setitem__(self, index: int, value: Any) -> None: ...
+    def __delitem__(self, key: int) -> None: ...
```
```suggestion
    def __getitem__(self, index: SupportsIndex | slice) -> Any: ...
    def __setitem__(self, index: SupportsIndex | slice, value: Any) -> None: ...
    def __delitem__(self, key: SupportsIndex | slice) -> None: ...
```
The `__<x>item__` dunders can also take `slice` objects or any object implementing the `__index__` methods (not just `int`).

BvB93 reviewed (changes requested):
Got a couple of suggestions, but things should be good to go afterwards.

MatthieuDartiailh commented:
I addressed all comments. Let me know if you want to see any other change.

MatthieuDartiailh committed:
MAINT: add missing dunder methods to nditer type hints

The presence of __getattr__ is not sufficient to enable all protocols so explicitely add the dunder methods supported by nditer

MatthieuDartiailh commented:
I messed up when addressing the comments so I had to squash a bunch of things and forced pushed. Sorry for the mess.

BvB93 reviewed (approved):
LGTM.

mattip commented:
No new tests or changes to existing tests are needed?

MatthieuDartiailh commented:
I can add a bunch of type checking related tests but I have no idea where they are located.

BvB93 commented:
See the [`ArrayTerator`](https://github.com/numpy/numpy/blob/092905febd53742d2a700322f88b652d6a5fd4fb/numpy/typing/tests/data/reveal/arrayterator.py) typing tests for a few examples; I'd suggest creating a new file specifically for `nditer`.

If you're willing to add tests, then the most import thing would be to check that the various legal statements 
return the expected type (directly calling the methods is less interesting to check here). For example:

``` python
import numpy as np

nditer_obj: np.nditer

with nditer_obj as context:
    reveal_type(context)   # E: numpy.nditer
```

MatthieuDartiailh commented:
Thanks. For ``__get/set/delitem__`, are you fine with simply implicitly invoking the protocol but not using any reveal since everything is Any for the time being ?

BvB93 commented:
> Thanks. For ``get/set/delitem``, are you fine with simply implicitly invoking the protocol but not using any reveal since everything is Any for the time being ?

I'd still use a `reveal_type` here. The actual revealed value is less interesting here (for now), 
but it does still allows for some basic verification.

MatthieuDartiailh committed:
first pass at adding nditer typing tests

MatthieuDartiailh commented:
I pushed some first tests for nditer however I am having a hard time running tests locally. Running `numpy.typing.test("slow", verbose=3)`, `test_reveal` are deselected and I could not really figure why. I would appreciate any pointer you may have.

MatthieuDartiailh commented:
Failures look unrelated to this work, I believe. Let me if I am wrong.

BvB93 commented:
> I pushed some first tests for nditer however I am having a hard time running tests locally. Running `numpy.typing.test("slow", verbose=3)`, test_reveal are deselected and I could not really figure why. I would appreciate any pointer you may have.

I usually run the `python ./runtests.py -t numpy/typing --mode=full` command from the command line.
I think running `numpy.typing.test("full") should also work.

BvB93 commented:
In any case, the added tests look good to me.

mattip commented:
Thanks @MatthieuDartiailh
